coordinator-common/src/main/java/org/apache/kafka/coordinator/common/runtime/PartitionWriter.java

```diff
@@ -107,4 +107,16 @@ CompletableFuture<VerificationGuard> maybeStartTransactionVerification(
         short producerEpoch,
         short apiVersion
     ) throws KafkaException;
+
+    /**
+     * Delete records from a topic partition until specified offset
+     * @param tp                The partition to delete records from
+     * @param deleteUntilOffset Offset to delete until, starting from the beginning
```

Comment:
I suggest `deleteBeforeOffset` rather than `deleteUntilOffset`. The latter is clearly non-inclusive, while the latter is a bit more ambiguous. I think the effect we want here is that if I provide offset 10, then offsets up to and including 9 may be deleted, but not 10.